### DISCUSSION TITLE
Is there a way to toggle lazygit instead of closing and opening it from beginning?

### DISCUSSION THREAD

**Discussion by shauryagoel** [LazyVim/LazyVim#4681] [Q&A] [answered]
If I open a lazygit terminal window and close it using `q`, the window is closed completely. Is there a way to hide it such that whenever I try to open lazygit again, it resumes the session?
I tried using [this](https://github.com/kdheepak/lazygit.nvim/issues/106#issuecomment-2239265186) comment's code and adjusting it to lazyvim, but, couldn't make it work.

**Comment by vdegenne** [comment]
If you close it using `ctrl-/` it closes it but keeps the state in the background I believe.

**Reply by shauryagoel** [reply]
I tried that, but, it is opening terminal after using `ctrl-/`.

**Reply by vdegenne** [reply]
You have to open lazygit using `<leader>gg`.
`ctrl-/` can be used to close it and retain the state.

That's more like a trick than a real solution though, personally I'm also not fully satisfied about various keys there, for instance having to use `q` instead of `escape` key to close lazygit is a bummer, and I agree using `<leader>gg` should probably reopen lazygit with saved state, maybe you can open an issue to ask for that.

**Reply by dpetka2001** [reply]
`q` is the default keymap for lazygit to exit. You can see this in [lazygit keymaps](https://github.com/nvim-telescope/telescope.nvim?tab=readme-ov-file#default-mappings). You can change that in your `config.yml` in your lazygit configuration directory. Nothing to do with LazyVim really. You should look if lazygit supports restoring a session first or some sort of toggle. Because LazyVim just opens lazygit in a Neovim terminal, sets some colors and that's it. The rest is lazygit behavior.

If lazygit doesn't support sessions, then your only option is toggling the terminal window with `<C-/>` as already mentioned, since that won't exit the terminal completely.

**Reply by shauryagoel** [reply]
@dpetka2001 Is there a way to toggle only the lazygit terminal window with `q` instead of `<C-/>`? I am already using lazygit from terminal and pressing `q` to quit has become a habit and I want to keep using it as well.

**Reply by dpetka2001** [reply]
lazygit `q` from terminal only exits, it doesn't store session. When you are in lazygit in a Neovim terminal window, then `q` from lazygit will take precedence over `q` from your mappings. You would have to change `q` from lazygit in your `config.yml` to something else so that the keys don't conflict if you want to use `q` to toggle the Neovim terminal, otherwise even if you use `q` to toggle the terminal when you're inside lazygit its own `q` which will exit will take precedence.

**Reply by shauryagoel** [reply]
But, wouldn't that require changing something in neovim config (or lazygit config inside neovim)? As I want to remove the mapping of `<C-/>` for lazygit only.
I remapped quit to some other key than `q` (in `config.yaml`). Now, opening lazygit inside neovim and pressing `q` does not do anything.

**Reply by dpetka2001** [reply]
You can set in your keymaps `vim.keymap.set("t", "q", "<cmd>close<cr>")` so that it will toggle off the terminal window.

**Reply by shauryagoel** [reply]
Yes, that works!!
However, is there a way to apply this only in the lazygit terminal window? Currently, even in a normal terminal window, if I type `q`, the window closes as well instead of typing the letter `q`.

**Reply by dpetka2001** [reply, marked as answer]
Try the following in your `~/.config/nvim/lua/config/autocmds.lua`
```lua
vim.api.nvim_create_autocmd("TermOpen", {
  pattern = "*",
  callback = function()
    local term_title = vim.b.term_title
    if term_title and term_title:match("lazygit") then
      -- Create lazygit specific mappings
      vim.keymap.set("t", "q", "<cmd>close<cr>", { buffer = true })
    end
  end,
})
```
Hope it's better than before.

**Reply by shauryagoel** [reply]
Yes, that works as expected. Thank you so much.

**Reply by dpetka2001** [reply]
It seems it's not needed to change the lazygit keymap in `config.yml` from my little testing. Check it also yourself, so if you use lazygit in your terminal from outside Neovim you can still use `q` to close it.

**Reply by dpetka2001** [reply]
@shauryagoel Please also check my previous comment. Maybe you don't need the change in lazygit's `config.yml` after all.

**Reply by shauryagoel** [reply]
Yes, when you mentioned changes required in lazygit config, I got a little confused as well because as per my understanding, only neovim config changes were required. 
Finally, I had to add only the autocmd that you mentioned and it solved my original question as well.
Thanks for the help.

**Reply by dpetka2001** [reply]
Yes that was my initial impression. But when i tested it with autocmd it wasn't the case. Sorry for confusing you, i also had a wrong understanding about it.

**Reply by vdegenne** [reply]
> ```lua
> vim.api.nvim_create_autocmd("TermOpen", {
>   pattern = "*",
>   callback = function()
>     local term_title = vim.b.term_title
>     if term_title and term_title:match("lazygit") then
>       -- Create lazygit specific mappings
>       vim.keymap.set("t", "q", "<cmd>close<cr>", { buffer = true })
>     end
>   end,
> })
> ```

I am taking that too, super useful :)

**Reply by folke** [reply]
Just so you know, that `TermOpen` keymap, will prevent you from typing `q` in a commit message for example, so I would not recommend it! 😅

You can hide lazygit with `<c-/>` if you want, then when opening again with `<leader>gg` it will show the previously hidden view.

Not needed to change anything for this to work.

**Reply by folke** [reply]
Also, the proper way to update the lazygit win config would be through the `snacks.nvim` options.
You can add `opts.styles.lazygit.keys.q = ...`

But again, would not recommend mapping `q` in terminal mode.

**Reply by dpetka2001** [reply]
Yeah that recommendation was before the introduction of `snacks.nvim` and I also hadn't tested how the commit message behaves, but it makes sense. Better do what @folke recommends, unless you don't mind not being able to type `q` :stuck_out_tongue:

**Reply by vdegenne** [reply]
Oh yes right, thanks for the info! I don't mind `<c-/>`

**Reply by fredrikaverpil** [reply]
I'm not using LazyVim but I'm now using either `<C-/>` to hide or `q` to fully quit.

```lua
function M.setup_terminal_keymaps()
  -- Both <C-/> and <C-_> are mapped due to the way control characters are interpreted by terminal emulators.
  -- ASCII value of '/' is 47, and of '_' is 95. When <C-/> is pressed, the terminal sends (47 - 64) which wraps around to 111 ('o').
  -- When <C-_> is pressed, the terminal sends (95 - 64) which is 31. Hence, both key combinations need to be mapped.

  -- <C-/> toggles the floating terminal
  local ctrl_slash = "<C-/>"
  local ctrl_underscore = "<C-_>"

  -- <C-A-/> toggles the split terminal
  local ctrl_alt_slash = "<C-A-/>"
  local ctrl_alt_underscore = "<C-A-_>"

  -- <Esc><Esc> in terminal mode sends <C-\><C-n> to exit terminal mode, see :h term
  vim.api.nvim_set_keymap("t", "<Esc><Esc>", "<C-\\><C-n>", { noremap = true })

  local floating_term_cmd = function()
    local cmd = { "zsh" }
    local opts = { cwd = vim.fn.getcwd() }
    Snacks.terminal.toggle(cmd, opts)
  end

  return {
    { ctrl_alt_slash, require("fredrik.utils.terminal").toggle_split_terminal, mode = { "n", "i", "t", "v" }, desc = "Toggle split terminal" },
    { ctrl_alt_underscore, require("fredrik.utils.terminal").toggle_split_terminal, mode = { "n", "i", "t", "v" }, desc = "Toggle split terminal" },

    { ctrl_slash, floating_term_cmd, mode = { "n", "i", "t", "v" }, desc = "Toggle floating terminal" },
    { ctrl_underscore, floating_term_cmd, mode = { "n", "i", "t", "v" }, desc = "Toggle floating terminal" },

    { ctrl_slash, "<cmd>close<cr>", mode = { "t" }, desc = "Hide Terminal" },
    { ctrl_underscore, "<cmd>close<cr>", mode = { "t" }, desc = "which_key_ignore" },
  }
end
```

Full source: https://github.com/fredrikaverpil/dotfiles/blob/d7d587118b1da9e39fa0751dbe22c5330ae4a87c/nvim-fredrik/lua/fredrik/config/keymaps.lua#L887-L919

LazyVim keymaps for reference: https://github.com/LazyVim/LazyVim/blob/d1529f650fdd89cb620258bdeca5ed7b558420c7/lua/lazyvim/config/keymaps.lua#L174-L182

**Reply by dragonlobster** [reply]
> Try the following in your `~/.config/nvim/lua/config/autocmds.lua`
> 
> ```lua
> vim.api.nvim_create_autocmd("TermOpen", {
>   pattern = "*",
>   callback = function()
>     local term_title = vim.b.term_title
>     if term_title and term_title:match("lazygit") then
>       -- Create lazygit specific mappings
>       vim.keymap.set("t", "q", "<cmd>close<cr>", { buffer = true })
>     end
>   end,
> })
> ```
> 
> Hope it's better than before.

I had an alternative where i set the filetype of lazygit window to `lazygit` (i personally think it should have its own filetype by default, like `snacks_lazygit`, currently it shares the filetype with `snacks_terminal`):

```lua
...
styles = {
    lazygit = {
        bo = {
            filetype = "lazygit"
        }
    }
}
...
```

then set the keymap in `after/ftplugin/lazygit.lua`:

```
vim.keymap.set("t", "q", "<cmd>close<cr>", { buffer = true })
```

of course like folke said setting q would prevent you from typing q in commit message
